### DISCUSSION TITLE
Port forwarding stopped working on vscode, but works on browser

### DISCUSSION THREAD

**Discussion by Findeton** [community/community#68854] [Codespaces]
### Select Topic Area

Bug

### Body

I am on an M1 Mac with VS Code version 1.82.2 . I'm developing a project using codespaces and VS Code with some ports forwarded. If I open the codespace in the browser, then I can access the forwarded ports (8080/8090 for example). When I open the codespace in VS Code, port forwarding doesn't seem to work anymore. When I open http://127.0.0.1/8090 the browser now says `127.0.0.1 didn’t send any data. ERR_EMPTY_RESPONSE` on Chrome and `The connection to the server was reset while the page was loading.` on Firefox. And if I open the codespace back on the browser, I can again access the forwarded ports.

These are services run on a docker-compose, like `keycloak` and `graphql-engine`. If instead I run node/react on a terminal I can access the app on the forwarded port 3000. Also this was working perfectly until recently, and I updated VS Code around that time. I've also tried rebooting the laptop and I haven't recently installed any new apps.

**Comment by Findeton** [comment]
For example if I open this simple [react project](https://github.com/Findeton/react-codespaces) as a codespace, in the browser I can open the forwarded port 3001 and it works right away. Now I open this same codespace in VS Code and the forwarded port 3001 doesn't work at all.

**Comment by GLStephen** [comment]
We are seeing similar issues with multiple projects of ours. Hitting the internal port from the workspace terminal of the codespace works fine, app runs as expected. Hitting the external forwarded port is a no go.

**Comment by Rothalack** [comment]
I'm also having this issue.

Based on your initial description, it sounded like you could see it working when you opened the codespace directly in browser and hosted from there. I am not getting it to host properly there either. So it's not working for me in both VSCode and directly in browser.

**Comment by mikebronner** [comment]
seeing this as well, opened a support ticket: https://support.github.com/ticket/personal/0/2357057

**Comment by edulix** [comment]
Related: https://github.com/orgs/community/discussions/68324

**Comment by ZeyadAzzam** [comment]
I'm also having the same issue :(

**Comment by nottisa** [comment]
Given this example it looks as If you are trying to go to a url on port 80 localhost, not actually successfully going to the port. The correct way to get to a port on localhost would be with localhost:port, also if it is forwarded you would need to use the forwarding URL given. Codespaces seems to have issues forwarding ports in general though... it also seems to be intermittent. Hopefully this helps...

**Comment by osortega** [comment]
@Findeton 
Sorry for the friction you've been having.
Do you mind sharing a codespace name that presented this issue so I could take a look?

At the same time if you could file a GitHub support ticket it would be great so that we can get more detailed logs from you.

**Comment by Findeton** [comment]
@osortega sure, a codespace name that is having this issue is "vigilant fishstick". If you go into port 8090 it will open keycloak, when you have the codespace open in the browser, but when it's in vscode it will not work.

I'll create a ticket later and let you know. Thanks for the interest, I hope this issue is solved soon enough...

**Reply by osortega** [reply]
Please when you open the ticket make sure to include the logs.
You can do that by first reproducing the issue and then exporting the logs with the command `Codespaces: Export Logs` and also if you could include the logs in `~/.vscode-remote/data/logs` from within the codespace, it would be really helpful.

**Comment by osortega** [comment]
@Findeton I took a look at telemetry and I want to ask you a couple of follow up questions:

When you say port forwarding doesn't work anymore when you open the Codespace in VS Code desktop, do you mean that the port stops being forwarded or do you mean that you try forwarding the port again and it fails?

If the issue is the former, what happens if you re-forward the port through VS Code desktop `Ports` tab?
If the issue is the latter, do you see any error message showing up in the VS Code window?

**Comment by Findeton** [comment]
@osortega what I mean is that in the VS Code ports tab, the port 8090 is there as if it was being correctly forwarded:

<img width="1025" alt="Screenshot 2023-10-03 at 19 37 37" src="https://github.com/community/community/assets/3913223/f725d664-90a6-4729-912c-3bd44f443004">


 But when I enter the address localhost:8090, it doesn't load. The error depends on the browser. I get this on Safari:

<img width="786" alt="Screenshot 2023-10-03 at 19 34 54" src="https://github.com/community/community/assets/3913223/4d83dc21-38f5-4979-9ed5-e03b06e5a1af">

This on Chrome (mac):

<img width="787" alt="Screenshot 2023-10-03 at 19 35 20" src="https://github.com/community/community/assets/3913223/998fc0b7-fb77-4db3-a98d-034896688021">

This on firefox (sorry for the Spanish):
<img width="670" alt="Screenshot 2023-10-03 at 19 36 20" src="https://github.com/community/community/assets/3913223/e60e6bc5-0417-49ce-bc51-cc7640d8b596">

And when I open that port with VS Code in the browser... it works.

**Comment by osortega** [comment]
@Findeton Ah I see, what happens if you stop forwarding the port and forward it again? Does it let you connect to it?

**Comment by Findeton** [comment]
@osortega  If I remove the port from the list of forwarded ports in the ports tab on VS code and I add it back... nothing happens, it doesn't work yet.

**Comment by osortega** [comment]
@Findeton I see, it would be really useful to look into detailed logs. Let me know when you file the support ticket so I can look into it.

Other option is that if you could provide a repro (with a public repository) and repro steps I could try the flow on my side and iterate faster on this.

**Comment by Findeton** [comment]
@osortega Fortunately I work on open source projects so you can try to open this repo as a codespace: https://github.com/sequentech/backend-services

**Reply by osortega** [reply]
@Findeton perfect! Mind providing repro steps?

**Reply by osortega** [reply]
@Findeton actually looks like the repository is locked for me.
Maybe providing a minimal repro would be good

**Reply by Findeton** [reply]
Sorry, I didn't notice this repo is not public yet. I'll create a simple version.

**Reply by Findeton** [reply]
@osortega you can use this repo https://github.com/Findeton/codespaces-port-error
Just open it as a codespace and when it's loaded, forward the port 8090. You'll see that it works on the browser (it will load keycloak) but not on VS Code.

**Reply by edulix** [reply]
I'd like to add that port forwarding in this example repo from @Findeton does work in VSCode... if it is forwarding a port launched from the command line instead from a docker container. And additionally, from a terminal, you can wget http://127.0.0.1:8090 and it works.

Because of the above, a workaround is to setup a nginx proxy run from a vscode task (or just launch it from the terminal manually), that redirects 8090 from container keycloak:8090 to localhost:8090, and then the port forwarding will work in VSCode. But I think this seems to then make port forwarding fail when launching the codespace in the browser. Tough luck.

And also, as other said, this seemed to start failing about 1 week ago. I tried to use an older VS Code version (1.81) but still didn't work to me. Also tried to downgrade some related extensions, still not working. But maybe I used the wrong versions.

**Reply by Findeton** [reply]
@osortega any updates on solving the issue? were you able to reproduce it? Also in Linux I noticed that when I'm not using codespaces (so, when I'm using devcontainers locally) I see the same issue, but if I have VS Code running, and I have a browser tab open trying to load the forwarded port `localhost:8090` (or whichever), it keeps trying to load. Now, the interesting part is that if I then CLOSE VS Code, the tab finally loads correctly. That doesn't help much because if I then try to reload the tab with VS Code closed it won't work.

**Reply by osortega** [reply]
@Findeton Yes, I'm able to reproduce this issue, I've been investigating this for a while but I have a suspicion that it might have to do with the repository configuration. I haven't been able to pinpoint where the problem is and why the browser is able to handle the requests.

Did this ever work for you on the repository you are working on? Or did something change recently in the `devcontainer` folder?

**Reply by Findeton** [reply]
@osortega Thanks for investigating it! Yes, this was working fine last week. and I'm  not aware of any changes. As far as I know, the same version of the repo was working last week and is not working anymore now (obviously in regards to port forwarding in VS Code).

**Reply by osortega** [reply]
@Findeton sorry for the delay. I've been investigating this and I narrowed down the issue to a few changes. Hopefully we can fix this soon but just so you know this is my priority and I will keep you posted.

**Reply by osortega** [reply]
@Findeton the issue should be fixed now. I verified with the repo https://github.com/Findeton/codespaces-port-error.
Again, sorry for the friction this caused and thank you for your patience!

**Reply by Findeton** [reply]
@osortega thanks for the fix! I confirm that port forwarding on VS Code with codespaces is working again.

**Comment by MohammedMSalha** [comment]
update VS Code !!

**Reply by Findeton** [reply]
I'm using version 1.83.0 on Mac, released yesterday, and it's still not working.

**Comment by GLStephen** [comment]
@osortega I have docker-in-docker loaded and when I tried to restart my nginx container, to see if it would forward the port, I got this error:

Error response from daemon: Cannot restart container xxx_xxxxx_nginx_1: driver failed programming external connectivity on endpoint xxx_xxxxx_nginx_1 (e43fbd83bf9805ac2f80e0a51ccc2f158d602af7dbe6b977e01c98dd6832f659): Error starting userland proxy: listen tcp4 0.0.0.0:8091: bind: address already in use

Around the same time this forwarding issue started we were getting similar errors as the above when doing full rebuilds on the workspace. During startup of the containers an error very similar to this would appear.

**Reply by osortega** [reply]
@GLStephen This seems like a different issue, but I'm happy to verify if you could provide me a way to reproduce this.
I'm curious to see if this has to do with the docker-compose configuration colliding with port forwarding somehow.

**Reply by GLStephen** [reply]
@osortega I agree it seems different, but the timing is more than suspicious. This repo and another similar one have been working this way for close to a year. Including supporting restarts of containers. There have not been any recent port changes and this issue started right around when forwarding stopped working in two repos and this rebuild issue started in those repos too. Unfortunately, none of our repos are public. Happy to try to work out some way for you to get more visibility.

**Reply by osortega** [reply]
@GLStephen The fix for the issue above just got released, do you mind double checking if your issue is fixed? If not please file a GitHub Support ticket so that we could get more details on this.

**Reply by GLStephen** [reply]
@osortega That also fixed the issue we were seeing. Both the port fowarding and the container restarts are working now.

**Reply by mikebronner** [reply]
I can confirm on our end that our CodeSpaces are now forwarding ports correctly again.

**Reply by osortega** [reply]
Awesome! Great to hear, thank you for verifying and sorry for the friction it caused.

**Comment by GLStephen** [comment]
@osortega Thank you very much for the fix. Will we get any sort of post mortem? I'd like to be able to explain internally expectations about bugs like this that interrupt dev workflow. Thanks again for the attention and resolution!

**Reply by mikebronner** [reply]
I would love to see a postmortem article as well. Management was not happy, and we had several days of reduced productivity that we need to explain. It questions the viability of CodeSpaces in their mind. (We are exploring various backup solutions for this for the next time we encounter an outage.)

**Reply by osortega** [reply]
@GLStephen @mikebronner Apologies for this issue. We take the reliability of Codespaces seriously and understand the impact bugs can cause to your productivity. Upon investigation, we found that a small fraction of users (<1%) experienced issues with port forwarding under a set of specific conditions related to port forwarding from the VS Code desktop client connecting to codespaces that use a docker-compose devcontainer configuration. This bug occurred after updating a set of packages on the host VM image. We've applied a mitigation and are improving our validation to ensure this doesn't happen again.

**Reply by mikebronner** [reply]
Thank you @osortega

**Reply by Mobeidat** [reply]
@osortega  - can we about the fix in detail? Thank you!
